Sources/_MatchingEngine/Regex/Parse/Mocking.swift

```diff
@@ -1,107 +1,142 @@
+    }
+  }
 }
 
+/// Drop a set of regex delimiters from the input string, returning the contents
+/// and the delimiter used. The input string must have valid delimiters.
+func droppingRegexDelimiters(_ str: String) -> (String, Delimiter) {
+  let utf8 = str.utf8
+  func stripDelimiter(_ delim: Delimiter) -> String? {
+    let prefix = delim.opening.utf8
+    let suffix = delim.closing.utf8
+    guard utf8.prefix(prefix.count).elementsEqual(prefix),
+          utf8.suffix(suffix.count).elementsEqual(suffix) else { return nil }
+
+    return String(utf8.dropFirst(prefix.count).dropLast(suffix.count))
+  }
+  for d in Delimiter.allCases {
+    if let contents = stripDelimiter(d) {
+      return (contents, d)
+    }
+  }
+  fatalError("No valid delimiters")
+}
+
+/// Attempt to lex a regex literal between `start` and `end`, returning either
+/// the contents and pointer from which to resume lexing, or an error.
 public func _lexRegex(
-  _ start: UnsafeRawPointer
-) throws -> (String, SyntaxOptions)? {
+  start: UnsafeRawPointer, end: UnsafeRawPointer
+) -> Result<(contents: String, Delimiter, end: UnsafeRawPointer), LexError> {
+  precondition(start <= end)
   var current = start
 
   func ascii(_ s: Unicode.Scalar) -> UInt8 {
     assert(s.value <= 0x7F)
     return UInt8(truncatingIfNeeded: s.value)
   }
-  func load(offset: Int) -> UInt8 {
-    current.load(fromByteOffset: offset, as: UInt8.self)
+  func load(offset: Int) -> UInt8? {
+    guard current + offset < end else { return nil }
+    return current.load(fromByteOffset: offset, as: UInt8.self)
   }
-  func load() -> UInt8 { load(offset: 0) }
-  func advance() { current = current.successor() }
-  func eat() -> UInt8 {
-    defer { advance() }
-    return load()
+  func load() -> UInt8? { load(offset: 0) }
+  func advance(_ n: Int = 1) {
+    precondition(current + n <= end, "Cannot advance past end")
+    current = current.advanced(by: n)
   }
 
-  guard eat() == ascii("'") else {
-    throw LexError.notRegexLiteralStart
+  func tryEat(_ utf8: String.UTF8View) -> Bool {
+    for (i, idx) in utf8.indices.enumerated() {
+      guard load(offset: i) == utf8[idx] else { return false }
+    }
+    advance(utf8.count)
+    return true
   }
 
-  let syntax: SyntaxOptions
-  let delimiter = eat()
-  switch delimiter {
-  case ascii("|"): syntax = .modern
-  case ascii("/"): syntax = .traditional
-  default: return nil
+  // Try to lex the opening delimiter.
+  guard let delimiter = Delimiter.allCases.first(
+    where: { tryEat($0.opening.utf8) }) else {
+    return .failure(LexError(.unknownDelimiter, resumeAt: current.successor()))
   }
 
   let contentsStart = current
-  let count: Int = try {
-    while true {
-      switch load() {
-      case 0: throw LexError.endOfString
-      case ascii("\\"):
-        // Skip next byte
-        advance()
+  while true {
+    switch load() {
+    case nil, ascii("\n"), ascii("\r"):
+      return .failure(LexError(.endOfString, resumeAt: current))
```

Comment:
Not for this PR, but how do you think we should do multi-line support? I guess with the version argument we can add that at any time

Reply:
Yeah, as long as we can pick it up by e.g a different delimiter, we could change the logic to support multiple lines

Reply:
The delimiter ideally would be something like 

```
let r = '''/
  regex
  '''/
```

or something.